crates/matrix-sdk/src/media.rs

```diff
@@ -165,53 +165,40 @@ impl Media {
         Self { client }
     }
 
-    /// Upload some media to the server.
-    ///
-    /// # Arguments
-    ///
-    /// * `content_type` - The type of the media, this will be used as the
-    ///   content-type header.
-    ///
-    /// * `data` - Vector of bytes to be uploaded to the server.
-    ///
-    /// * `request_config` - Optional request configuration for the HTTP client,
-    ///   overriding the default. If not provided, a reasonable timeout value is
-    ///   inferred.
+    /// Upload some media to the server using the provided
+    /// [`SendMediaUploadRequest`].
     ///
     /// # Examples
     ///
     /// ```no_run
     /// # use std::fs;
-    /// # use matrix_sdk::{Client, ruma::room_id};
+    /// # use matrix_sdk::{Client, ruma::room_id, SendMediaUploadRequest};
     /// # use url::Url;
     /// # use mime;
     /// # async {
     /// # let homeserver = Url::parse("http://localhost:8080")?;
     /// # let mut client = Client::new(homeserver).await?;
     /// let image = fs::read("/home/example/my-cat.jpg")?;
     ///
-    /// let response =
-    ///     client.media().upload(&mime::IMAGE_JPEG, image, None).await?;
+    /// let send_media_request = SendMediaUploadRequest::new(client.clone(), image)
+    ///     .with_content_type(mime::IMAGE_JPEG.essence_str().to_string())
+    ///     .with_filename(Some("my-cat.jpg"));
+    ///
+    /// let response = client.media().upload(send_media_request).await?;
     ///
     /// println!("Cat URI: {}", response.content_uri);
     /// # anyhow::Ok(()) };
     /// ```
-    pub fn upload(
+    pub async fn upload(
         &self,
-        content_type: &Mime,
-        data: Vec<u8>,
-        request_config: Option<RequestConfig>,
-    ) -> SendMediaUploadRequest {
-        let request_config = request_config.unwrap_or_else(|| {
-            self.client.request_config().timeout(Self::reasonable_upload_timeout(&data))
-        });
-
-        let request = assign!(media::create_content::v3::Request::new(data), {
-            content_type: Some(content_type.essence_str().to_owned()),
+        send_media_upload_request: SendMediaUploadRequest,
+    ) -> Result<media::create_content::v3::Response, Error> {
```

Comment:
This is not right. Instead, it should take the basic parameters that are always needed for a `SendMediaUploadRequest`, and return it. As a result, it should not be async, and it should resolve the response. The way to use it would be:

```
client.media().upload(file_bytes_vec).with_request_config(RequestConfig::default()).await?;
```

and so on.

It may be better to split this change into a separate PR? The history of this one is already a bit messy, and this is a problem orthogonal to the one at hand. (I'd be happy reviewing the split up PR too, if that helps speeding up the review!)
```